packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -585,12 +585,35 @@ abstract class Route<T> extends _RoutePlaceholder {
     return currentRouteEntry.route == this;
   }
 
+  bool _hasActiveParentRouteBelow(NavigatorState? parent) {
+    if (parent == null) {
+      return false;
+    }
+
+    final _History history = parent._history;
+    if (history.length > 1) {
+      for (final _RouteEntry entry in history) {
+        if (_RouteEntry.isPresentPredicate(entry)) {
+          return true;
+        }
+      }
+    }
+
+    return _hasActiveParentRouteBelow(parent._parentNavigator);
+  }
+
   /// Whether there is at least one active route underneath this route.
   @protected
   bool get hasActiveRouteBelow {
     if (_navigator == null) {
       return false;
     }
+
+    final bool hasActiveParentRouteBelow = _hasActiveParentRouteBelow(_navigator!._parentNavigator);
```

Comment:
in nested case, it seem weird that you tap the inner appbar but then the outer navigator gets pop. I think this behavior may have different expectation based on use case.

Reply:
If we have one structure like this: 
- MaterialApp
  - Navigator (MainShellRoute)
    - MainScreen
    - Navigator (AnotherShellRoute)
     - AnotherScreen

When AnotherScreen is popped, since his parent Navigator (AnotherShellRoute) don't have any child page it should be popped as well, no?

And what would you recommend to do in cases where we have multiple Navigators?
Today my work around is force adding a BackButton to every page that can't include it auto, which for me is not desirable.

Reply:
As a user I would expect the back button should only pop content in the area that it surrounds.  If `AnotherShellRoute` only taken part of the screen and there isn't a route in `AnotherShellRoute` to be popped, there shouldn't be a back button in `AnotherShellRoute`.

It would be a different story if MainScreen is just an empty container and `AnotherShellRoute` take up the whole screen though, but in that case why would you need the MainShellRoute in the first place?

Reply:
My use case AnotherShellRoute is used to add the NavigationRail, something like this:
<img width="830" alt="Screenshot 2024-05-16 at 10 06 34" src="https://github.com/flutter/flutter/assets/38299943/e4bf9ff2-d904-4b62-a42a-424af1a6cbf1">

other possible solution would be changing on go_router package, abstracting this rule on "MaterialPage/CupertinoPage" and Navigator just for the ShellRoute.

Reply:
I as a user feels a bit weird about if there is a back button on page1 and press it pop the entire page includes the sidebar, but I do understand people may have special UX requirement. In this case, I think you should have a special backbutton implement in your code base to handle this use case.

This is just my opinion though, so I may be wrong. but as of now, I am not confident moving forward with this pr since we are just 2 data points. We can revisit this if we have more people asking to change or preserve this behavior.

WDYT?

Reply:
Fine for me